### PR TITLE
[bug 1221295] Mobile activation snippet (widget look)

### PR DESCRIPTION
**Pull request by ckprice**
Based off an [existing template](https://github.com/mozilla/snippets/blob/6c428fbbf17c650332ba924a9f2dcb528ee4778e/campaigns/sms-android/android-email-sms-configure.html), with an updated look/feel to match the Send to Device widget seen on the Android [product page](https://www.mozilla.org/en-US/firefox/android/).

@glogiotatidis @schalkneethling r?

Also, per my email, I'd like to be able to use `Open Sans` on this widget. Is that possible?

### CODE REVIEW

**Comment by ckprice** [comment]
@glogiotatidis I tried using the font as prescribed [here](https://github.com/mozilla/snippets/blob/6c428fbbf17c650332ba924a9f2dcb528ee4778e/generic-templates/video-snippets/simple-over-search-bar.html), with no affect.

I borrowed from a [3rd party source](https://gist.github.com/nottrobin/5237540) temporarily. How best to get this into our infrastructure?

Hoping to get this one reviewed today if possible.

File: campaigns/mobile/mobile-snippet-widget.html [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+      background: #F8F8F8;
+      color: #56565a;
+      text-align: center;
+      border-radius: 10px;
+      font-family: open-sans-{{ snippet_id }};
+    }
+
+    #send-to-device-{{ snippet_id }} a {
+        color: #0c99d5;
+        -webkit-transition: color .1s ease-in-out;
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: This only runs in Fx so no need for the webkit line

File: campaigns/mobile/mobile-snippet-widget.html [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+      font-weight: normal;
+      font-style: normal;
+  }
+
+    #send-to-device-{{ snippet_id }} {
+      width: 100%;
+      margin: 0 auto;
+      background: #F8F8F8;
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Nit: `backgound-color`

File: campaigns/mobile/mobile-snippet-widget.html [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+      position: relative;
+      padding: 0 20px;
+    }
+    #send-to-device-form-{{ snippet_id }} .error-list {
+      margin-bottom: 20px;
+      text-align: left;
+    }
+
+    #send-to-device-form-{{ snippet_id }} .error-list li {
+      background: #c33b32;
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Nit: `backgound-color`

File: campaigns/mobile/mobile-snippet-widget.html [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+      float: right;
+      width: 100px;
+    }
+
+    #send-to-device-form-{{ snippet_id }} .input .button-flat {
+      padding: 7px 5px;
+      min-width: 100px;
+      margin-bottom: 0;
+      background-color: #0c99d5;
+      -webkit-transition: background-color .1s ease-in-out;
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Don't need the `-webkit-` prefixed rule

File: campaigns/mobile/mobile-snippet-widget.html [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+      transition: background-color .1s ease-in-out;
+      font-size: .875rem;
+      border: 0;
+      cursor: pointer;
+      background: #00a0dc;
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: `background-color`

File: campaigns/mobile/mobile-snippet-widget.html [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+      color: #fff;
+      text-align: center;
+      display: inline-block;
+      border-radius: 6px;
+    }
+
+     #send-to-device-form-{{ snippet_id }} .insensitive {
+       background-color: #9ca3aa !important;
+       cursor: default !important;
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Do we *need* the `!important` here?

File: campaigns/mobile/mobile-snippet-widget.html [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+     }
+
+    #send-to-device-form-{{ snippet_id }} .input #id-input {
+      width: 240px;
+      height: 32px;
+      padding: 5px 10px;
+      line-height: 2;
+      -webkit-transition: none;
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Not needed

File: campaigns/mobile/mobile-snippet-widget.html [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+
+    #send-to-device-form-{{ snippet_id }} .thank-you:before {
+        position: absolute;
+        top: 0;
+        left: 50%;
+        content: '';
+        width: 72px;
+        height: 72px;
+        margin-left: -36px;
+        background-image: url("https://www.mozilla.org/media/img/send-to-device/device-icon.28cef022bba2.png");
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: We probably want to convert this to a data url? Else perhaps use the CDN?

File: campaigns/mobile/mobile-snippet-widget.html [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+      border-bottom-left-radius: 10px;
+      border-bottom-right-radius: 10px;
+    }
+
+    #send-to-device-{{ snippet_id }} footer {
+      margin-top: 20px;
+      padding: 10px 0;
+      background: #e1e2e2;
+      border-top: 1px solid #d3d4d4;
+      zoom: 1;
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Do we need `zoom: 1;` for Fx?

File: campaigns/mobile/mobile-snippet-widget.html [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+               input_type = checkInput(input.value);
+               if (input_type) {
+                   button.disabled = false;
+                   button.classList.remove('insensitive');
+               }
+               else {
+                   button.disabled = true;
+                   button.classList.add('insensitive');
+               }
+               console.log('checking ' + input.value + ' = ' + input_type);
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Left over debugging?

File: campaigns/mobile/mobile-snippet-widget.html [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+           });
+
+          if ("createEvent" in document) {
+              var evt = document.createEvent("HTMLEvents");
+              evt.initEvent("input", false, true);
+              input.dispatchEvent(evt);
+          }
+          else
+              input.fireEvent("input");
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: We do not need to check support here, we can just go ahead and use `createEvent`

File: campaigns/mobile/mobile-snippet-widget.html [hunk 12]
```diff
@@ -0,0 +1,420 @@
...
+                } else if (check_phone && '{{ include_sms }}' === 'True') {
+                    return 'phone';
+                } else {
+                    return false;
+                }
+           }
+
+           function sendSMSToBasket () {
+                var url = 'https://basket.mozilla.org/news/subscribe_sms/';
+                var r = new XMLHttpRequest();
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Nit: `r` is not a great variable name, can we use `request` instead?

File: campaigns/mobile/mobile-snippet-widget.html [hunk 13]
```diff
@@ -0,0 +1,420 @@
...
+           }
+
+           function changeState (new_state) {
+              if (new_state === 'success') {
+                document.getElementById('welcome-header').style.display = 'none';
+                document.getElementById('thank-you-message').style.display = 'block';
+                document.getElementById('form-input').style.display = 'none';
+                document.getElementById('thank-you-sub-message').style.display = 'block';
+                document.getElementById('error').style.display = 'none';
+              }else if (new_state === 'error') {
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Nit: space after `}`

File: campaigns/mobile/mobile-snippet-widget.html [hunk 14]
```diff
@@ -0,0 +1,420 @@
...
+                document.getElementById('welcome-header').style.display = 'block';
+                document.getElementById('thank-you-message').style.display = 'none';
+                document.getElementById('form-input').style.display = 'block';
+                document.getElementById('thank-you-sub-message').style.display = 'none';
+                document.getElementById('error').style.display = 'block';
+              }else {
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Nit: space after `}`

File: campaigns/mobile/mobile-snippet-widget.html [hunk 15]
```diff
@@ -0,0 +1,420 @@
...
+    </div>
+    <footer>
+      <ul>
+        <li class="store">
+          {{ store_footer|safe }}
+        </li>
+      </ul>
+    </footer>
+  </section>
+</div>
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Missing eof empty line

File: campaigns/mobile/mobile-snippet-widget.html [hunk 16]
```diff
@@ -0,0 +1,420 @@
...
+                document.getElementById('thank-you-message').style.display = 'none';
+                document.getElementById('form-input').style.display = 'block';
+                document.getElementById('thank-you-sub-message').style.display = 'none';
+                document.getElementById('error').style.display = 'none';
+              }
+           }
+       });
+   })();
+   //]]>
+  </script>
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: Should we move the JS to after the HTML?

File: campaigns/mobile/mobile-snippet-widget.html [hunk 17]
```diff
@@ -0,0 +1,420 @@
...
+            </div>
+          </div>
+          <div class="legal">
+            {{ terms_footer|safe }}
+          </div>
+        </div>
+        <div class="thank-you hidden" id="thank-you-sub-message">
+          <div>{{ thank_you_sub_message }}</div>
+        </div>
+        <div class="loading-spinner"></div>
```
Comment (schalkneethling) [inline campaigns/mobile/mobile-snippet-widget.html]: We do not use this anywhere so can safely remove.

**Comment by schalkneethling** [comment]
@ckprice Do you have some sample copy and assets that I can use to test the snippet? I can just make up stuff but, probably a better idea to use something that actually matches how this will be used in the wild. Thanks!

**Comment by glogiotatidis** [comment]
> @glogiotatidis I tried using the font as prescribed here, with no affect.
OK the font in that snippet is probably not open sans anyway. I just linked for the example.

> I borrowed from a 3rd party source temporarily. How best to get this into our infrastructure?
Copy pasting that code should be enough. Doesn't that work for you?

**Comment by ckprice** [comment]
@schalkneethling okay, comments addressed! Please have a look.

@glogiotatidis re: font, of course. Copy/paste works fine thanks!

**Comment by ckprice** [comment]
>@ckprice Do you have some sample copy and assets that I can use to test the snippet? I can just make up stuff but, probably a better idea to use something that actually matches how this will be used in the wild. Thanks!

Here is the snippet on staging https://snippets.allizom.org/show/98/ (may need to ask Giorgos for access).

**Comment by schalkneethling** [comment]
Looks great! r+ :v:

File: campaigns/mobile/mobile-snippet-widget.html [hunk 18]
```diff
@@ -0,0 +1,405 @@
...
+                  }
+              }
+              request.open('POST', url);
+              request.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
+              request.send(params);
+         }
+
+         function sendEmailToBasket () {
+              //TODO: get language from data-locale
+              var lang = '{{ language }}';
```
Comment (glogiotatidis) [inline campaigns/mobile/mobile-snippet-widget.html]: `document.getElementById('snippet_set').getAttribute('data-locale');`

Comment (ckprice) [inline campaigns/mobile/mobile-snippet-widget.html]: Thanks Giorgos. We are adding language as part of the template for ease of
testing other languages for the snippet admins. I'll have to come up with a
better way in the future, and will note this.

On Thursday, November 12, 2015, Giorgos Logiotatidis <
notifications@github.com> wrote:

> In campaigns/mobile/mobile-snippet-widget.html
> <https://github.com/mozilla/snippets/pull/60#discussion_r44665969>:
>
> > +
> > +              var params = 'mobile_number=' + input.value + '&msg_name={{ sms_id }}';
> > +
> > +              request.onreadystatechange = function() {
> > +                  if (request.readyState == 4) {
> > +                    changeState(request.status);
> > +                  }
> > +              }
> > +              request.open('POST', url);
> > +              request.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
> > +              request.send(params);
> > +         }
> > +
> > +         function sendEmailToBasket () {
> > +              //TODO: get language from data-locale
> > +              var lang = '{{ language }}';
>
> document.getElementById('snippet_set').getAttribute('data-locale');
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/snippets/pull/60/files#r44665969>.
>


-- 
Cory Price
/ckprice

File: campaigns/mobile/mobile-snippet-widget.html [hunk 19]
```diff
@@ -0,0 +1,405 @@
...
+
+    #send-to-device-{{ snippet_id }}.logo .form-container:before {
+        position: absolute;
+        top: -50px;
+        left: 50%;
+        width: 80px;
+        height: 80px;
+        margin-left: -40px;
+        content: '';
+        background-image: url({{ image }});
```
Comment (glogiotatidis) [inline campaigns/mobile/mobile-snippet-widget.html]: Add quotes `url('{{ image }}')`

File: campaigns/mobile/mobile-snippet-widget.html [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+          <li>{{ error_message }}</li>
+        </ul>
+        <ul class="error-list hidden" id="error-toomany">
+          <li>Too many requests</li>
+        </ul>
+        <div class="input" id="form-input">
+          <div class="inline-field">
+            <label id="form-input-label" for="id-input">{{ sub_header }}</label>
+            <div class="form-input">
+              <input id="id-input" name="phone-or-email" type="text" class="number" required placeholder="{{ input_placeholder }}">
```
Comment (glogiotatidis) [inline campaigns/mobile/mobile-snippet-widget.html]: Please end with `/>` for valid XML

**Comment by ckprice** [comment]
@glogiotatidis @schalkneethling thanks! Last two Giorgos comments addressed and squashed.

**Comment by ckprice** [comment]
@glogiotatidis really hoping you can give this a quick review and push to production tomorrow your time AM if possible. See https://bugzilla.mozilla.org/show_bug.cgi?id=1224407 for more detail.

**Comment by glogiotatidis** [comment]
Because the sms send request take a few seconds, you can set button back to inactive after user clicks to avoid multiple sends.

**Comment by glogiotatidis** [comment]
I'm merging this one. Cory feel free to send a new one if you want to tackle my last comment.

Thanks!

**Pull request merged**